derive/src/lib.rs

```diff
@@ -184,6 +184,31 @@ macro_rules! expr {
     }}
 }
 
+fn gen_header(
+    generic_header: syn::ImplGenerics,
+    ident: syn::Ident,
+    input_ident: syn::Ident,
+    generic: syn::TypeGenerics,
+    ty: &syn::Type,
+) -> proc_macro::TokenStream {
```

Comment:
```suggestion
) -> syn::Item {
```

The `item` macro is similar to `quote`, but it auto-parses into a `syn::Item` for convenience.

Reply:
I thought just doing that change as suggested will again error asking for `proc_macro::TokenStream`! For me it starts a diagnostic feedback loop: errors for `syn::Item`, flips and errors for `proc_macro::TokenStream`, flips and errors for `syn::Item`, ...

I spent a while trying to debug this using `From` and `Into` traits in `syn` and `proc_macros` to try to enforce one return type without any luck.

Reply:
Oh, that's really strange. I'll check it out locally

Reply:
I should have been more clear posting the PR, I was mostly wondering if wrapping the macro in a function was the same approach you were thinking about.

Reply:
Here is the other error:
```zsh
Compiling gfp-derive v0.1.0 (/media/oliver/WD/generic-field-projection/derive)
error[E0308]: mismatched types
   --> derive/src/lib.rs:194:5
    |
193 |   ) -> syn::Item {
    |        --------- expected `syn::Item` because of return type
194 | /     item!(
195 | |         unsafe impl #generic_header ::gfp_core::Field for #ident<super::#input_ident #generic> {
196 | |             type Parent = super::#input_ident #generic;
197 | |             type Type = #ty;
...   |
208 | |         }
209 | |     )
    | |_____^ expected enum `syn::Item`, found struct `proc_macro::TokenStream`
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```
This 'seems closer'(tm) since the function calls in the module constructor now aren't erroring but I wasn't able to debug this.

Reply:
So, if you compile with the environment variable `RUSTFLAGS="-Z macro-backtrace"` on nightly, you'll see the following error:
```rust
error[E0308]: mismatched types
   --> C:\Users\Krishna\.cargo\registry\src\github.com-1ecc6299db9ec823\syn-1.0.60\src\parse_macro_input.rs:113:24
    |
108 | / macro_rules! parse_macro_input {
109 | |     ($tokenstream:ident as $ty:ty) => {
110 | |         match $crate::parse_macro_input::parse::<$ty>($tokenstream) {
111 | |             $crate::__private::Ok(data) => data,
112 | |             $crate::__private::Err(err) => {
113 | |                 return $crate::__private::TokenStream::from(err.to_compile_error());
    | |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `syn::Item`, found struct `proc_macro::TokenStream`
...   |
127 | |     };
128 | | }
    | |_- in this expansion of `syn::parse_macro_input!` (#2)
    |
   ::: derive\src\lib.rs:173:1
    |
173 | / macro_rules! item {
174 | |     ($($tokens:tt)*) => {{
175 | |         let quote = TokenStream::from(quote!($($tokens)*));
176 | |         syn::parse_macro_input!(quote as syn::Item)
    | |         ------------------------------------------- in this macro invocation (#2)
177 | |     }}
178 | | }
    | |_- in this expansion of `item!` (#1)
...
194 |       let () = item!(
    |  ______________-
195 | |         unsafe impl #generic_header ::gfp_core::Field for #ident<super::#input_ident #generic> {
196 | |             type Parent = super::#input_ident #generic;
197 | |             type Type = #ty;
...   |
208 | |         }
209 | |     );
    | |_____- in this macro invocation (#1)
```

What's happening is that `syn::parse_macro_input` is assuming that it's in a function that returns a `TokenStream`, so it's trying to return an error on failed parse. (Of course this won't happen in our case, but this is handling a more general case).    
Can you change the `item` macro to use `syn::parse::<syn::Item>($macro_input).unwrap()` (and possibly other similar macros in similar ways)? This way it doesn't try to return a `TokenStream`.

Reply:
That works thank you!!